### PR TITLE
Added Oauth2 functionality for REST APIs

### LINKED ISSUE #934: [Feature] Add OAuth2 authentication for API integrations
## Summary
Users need a way to configure OAuth authentication for their APIs. We can have a section where users can configure the OAuth2 clientId / secret in the API pane. 

## Motivation
Most SAAS integrations require OAuth and most backends support this as well.

## Additional Context
The OAuth token should be fetched, stored, and refreshed by the server using the client-Id/secret. The token should not be fetched on every action execution. The token should be auto refreshed

### PR DESCRIPTION
**Pull request by nidhi-nair**
OAuth2 is inherently part of the API plugin. The following functionality has been added to the plugin as part of this PR:

1. Client credentials flow for OAuth2. Users can now connect to APIs that require long-lived or short lived access tokens. Since the automated flow for API plugins cannot be interrupted, this change only supports 2LO. 3LO will remain a front end feature for now.
2. All client secrets are encrypted on the backend. This also means that they are not editable once configured. Users will need to re-enter their authentication details.
3. Customizable Header prefix (defaults to Bearer)
4. Choice to include access token in either header or query params (Official documentation recommends using headers)
5. Auto refresh of credentials on expiry. The client credentials flow typically requests for a new token on every request. However, for cases where the token response supports an expiry datetime instant, this change allows token reuse. The relevant fields that are tracked are:
- expires_at
- issued_at and expires_in
6. Configurable scopes
7. Extensible OAuth2 type to allow customizations over the default OAuth2 behaviour
8. Authentication types can be swapped out in case an API configuration needs to be changed. The request can simply start using the new type to do so. Note that previous configurations will not be saved.
 
Refer to [this Notion document](https://www.notion.so/appsmith/Datasource-and-plugin-execution-related-changes-8282acbd4e334bc5aea6a757bc669ccd#c1733db392ea42a2b03f52ca8e0c0d60) for context

Fixes #934

## Type of change
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update

## How Has This Been Tested?
- JUnit for edge cases
- Manually tested on local since credentials cannot be shared on JUnit

*Creating datasource of OAuth2 type*
image

image

*Creating action for this datasource*
image
image

*Executing action (2LO against Facebook API)*
image
image

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/connections/OAuth2Connection.java [hunk 1]
```diff
@@ -1,28 +1,174 @@
...
+    }
 
-        return null;
+    private Mono<OAuth2> generateOAuth2Token(OAuth2 oAuth2) {
+        // Webclient
+        WebClient webClient = WebClient.builder()
+                .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
+                .exchangeStrategies(ExchangeStrategies
+                        .builder()
+                        .codecs(configurer -> configurer.defaultCodecs().maxInMemorySize(/* 10MB */ 10 * 1024 * 1024))
```
Comment (trishaanand) [inline app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/connections/OAuth2Connection.java:77]: Could we define the memory size as a variable defined at the top of the class for easy grep and update later?

File: app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/connections/OAuth2ConnectionTest.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+import org.junit.Test;
+
+import java.time.Duration;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class OAuth2ConnectionTest {
+
+    @Test
+    public void testNullConnection() {
```
Comment (trishaanand) [inline app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/connections/OAuth2ConnectionTest.java:12]: Could we add more test cases using mocks?

File: app/server/appsmith-plugins/restApiPlugin/pom.xml [hunk 3]
```diff
@@ -103,6 +103,28 @@
...
 
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <!-- https://mvnrepository.com/artifact/org.powermock/powermock-module-junit4 -->
+        <dependency>
+            <groupId>org.powermock</groupId>
```
Comment (trishaanand) [inline app/server/appsmith-plugins/restApiPlugin/pom.xml:114]: Could we continue to use Mockito which we use for the server tests?

Comment (nidhi-nair) [reply app/server/appsmith-plugins/restApiPlugin/pom.xml:114]: Powermock is something that adds functionality on top of Mockito, in this case, to test private and static methods. Mockito does not support this in its vanilla form. I can remove it if we figure that such tests can not be performed.

**Review by trishaanand — APPROVED** [review]

**Pull request merged**